components/hab/src/command/pkg/upload.rs

```diff
@@ -173,29 +185,43 @@ async fn upload_into_depot(ui: &mut UI,
     ui.status(Status::Uploaded, ident)?;
 
     // Promote to additional_release_channel if specified
-    if package_uploaded && additional_release_channel.is_some() {
+    if package_exists_in_target && additional_release_channel.is_some() {
         let channel = additional_release_channel.clone().unwrap();
-        ui.begin(format!("Promoting {} to channel '{}'", ident, channel))?;
-
-        if channel != ChannelIdent::stable() && channel != ChannelIdent::unstable() {
-            match api_client.create_channel(&ident.origin, &channel, token)
-                            .await
-            {
-                Ok(_) => (),
-                Err(api_client::Error::APIError(StatusCode::CONFLICT, _)) => (),
-                Err(e) => return Err(Error::from(e)),
-            };
+        match promote_to_channel(ui, api_client, (ident, target), channel, token).await {
+            Ok(_) => (),
+            Err(e) => return Err(e),
         }
+    }
+
+    Ok(())
+}
 
-        match api_client.promote_package((ident, target), &channel, token)
+async fn promote_to_channel(ui: &mut UI,
+                            api_client: &BuilderAPIClient,
+                            (ident, target): (&PackageIdent, PackageTarget),
+                            channel: ChannelIdent,
+                            token: &str)
+                            -> Result<()> {
+    ui.begin(format!("Promoting {} to channel '{}'", ident, channel))?;
+
+    if channel != ChannelIdent::stable() && channel != ChannelIdent::unstable() {
+        match api_client.create_channel(&ident.origin, &channel, token)
                         .await
         {
             Ok(_) => (),
+            Err(api_client::Error::APIError(StatusCode::CONFLICT, _)) => (),
             Err(e) => return Err(Error::from(e)),
         };
-        ui.status(Status::Promoted, ident)?;
     }
 
+    match api_client.promote_package((ident, target), &channel, token)
+                    .await
+    {
+        Ok(_) => (),
+        Err(e) => return Err(Error::from(e)),
+    };
```

Comment:
`?` operator?

Reply:
Let's do it! Addressed in https://github.com/habitat-sh/habitat/pull/7768/commits/9f75668f9a777642aae2b820bab1cd280fd00898